src/components/views/elements/MessageEditor.js

```diff
@@ -0,0 +1,190 @@
+        // modelOutput.textContent = JSON.stringify(this.model.serializeParts(), undefined, 2);
+    }
+
+    _onInput = (event) => {
+        const sel = document.getSelection();
+        // console.log("finding newValue", this._editorRef.innerHTML, sel);
+        const {caret, text} = getCaretOffsetAndText(this._editorRef, sel);
+        this.model.update(text, event.inputType, caret);
+    }
+
+    _onKeyDown = (event) => {
+        if (event.metaKey || event.altKey || event.shiftKey) {
+            return;
+        }
+        if (!this.model.autoComplete) {
+            return;
+        }
+        const autoComplete = this.model.autoComplete;
+        switch (event.key) {
+            case "Enter":
+                autoComplete.onEnter(event); break;
+            case "ArrowUp":
+                autoComplete.onUpArrow(event); break;
+            case "ArrowDown":
+                autoComplete.onDownArrow(event); break;
+            case "Tab":
+                autoComplete.onTab(event); break;
+            case "Escape":
+                autoComplete.onEscape(event); break;
+            default:
+                return; // don't preventDefault on anything else
+        }
+        event.preventDefault();
+    }
+
+    _onCancelClicked = () => {
+        dis.dispatch({action: "edit_event", event: null});
+    }
+
+    _onSaveClicked = () => {
+        const newContent = {
+            "msgtype": "m.text",
+            "body": textSerialize(this.model),
+        };
+        if (requiresHtml(this.model)) {
+            newContent.format = "org.matrix.custom.html";
+            newContent.formatted_body = htmlSerialize(this.model);
+        }
+        const content = Object.assign({
+            "m.new_content": newContent,
+            "m.relates_to": {
+                "rel_type": "m.replace",
+                "event_id": this.props.event.getOriginalId(),
+            },
+        }, newContent);
+
+        const roomId = this.props.event.getRoomId();
+        this.context.matrixClient.sendMessage(roomId, content);
+
+        dis.dispatch({action: "edit_event", event: null});
+    }
+
+    _collectEditorRef = (ref) => {
+        this._editorRef = ref;
+    }
+
+    _collectAutocompleteRef = (ref) => {
+        this._autocompleteRef = ref;
+    }
+
+    _onAutoCompleteConfirm = (completion) => {
+        this.model.autoComplete.onComponentConfirm(completion);
+    }
+
+    _onAutoCompleteSelectionChange = (completion) => {
+        this.model.autoComplete.onComponentSelectionChange(completion);
+    }
+
+    componentDidMount() {
+        this._updateEditorState();
+    }
+
+    render() {
+        let autoComplete;
+        if (this.state.autoComplete) {
+            const query = this.state.query;
+            const queryLen = query.length;
+            autoComplete = <div className="mx_MessageEditor_AutoCompleteWrapper">
+                <Autocomplete
+                    ref={this._collectAutocompleteRef}
```

Comment:
fwiw this can be done inline as `ref={(ref) => this._autocompleteRef = ref}`

Reply:
Oh yes. I though the coding style doc said this was a perf problem, but it seems to allow it for refs as an exception :)